jax/experimental/export/export.py

```diff
@@ -1072,12 +1104,47 @@ def convert_shape(x: ir.Value, x_aval: core.AbstractValue, new_aval: core.Abstra
                              kept_args)
   # The ctx.avals_out already contain the abstract values refined by
   # _call_exported_abstract_eval.
-  return tuple(
+  results = tuple(
       convert_shape(out, out_aval, refined_out_aval)
       for out, out_aval, refined_out_aval in zip(call.results, exported.out_avals, ctx.avals_out))
+  # Apply out_shardings
+  results = tuple(
+    wrap_with_sharding(ctx, exported, x, x_aval, x_sharding)
+    for x, x_aval, x_sharding in zip(results, ctx.avals_out, exported.out_shardings)
+  )
+  return results
 
 
 for _p in ("cpu", "tpu", "cuda", "rocm"):
   mlir.register_lowering(call_exported_p,
                          functools.partial(_call_exported_lowering, platform=_p),
                          platform=_p)
+
+def wrap_with_sharding(ctx: mlir.LoweringRuleContext,
+                       exported: Exported,
+                       x: ir.Value,
+                       x_aval: core.AbstractValue,
+                       x_sharding: Sharding) -> ir.Value:
+  if sharding_impls.is_unspecified(x_sharding):
+    return x
+  if not isinstance(x_sharding, sharding_impls.GSPMDSharding):
+    raise NotImplementedError(
+      f"call_exported implemented only for GSPMDSharding. Found {x_sharding}"
+    )
+  axis_context = ctx.module_context.axis_context
+  if isinstance(axis_context, sharding_impls.ShardingContext):
+    ctx_device_assignment = axis_context.device_assignment
+  elif isinstance(axis_context, sharding_impls.SPMDAxisContext):
+    ctx_device_assignment = list(axis_context.mesh.devices.flat)
+  else:
+    raise NotImplementedError(type(axis_context))
+  sharding_device_assignment = x_sharding._devices
+  # TODO: should we check that the device assignment has the same ids?
+  if len(ctx_device_assignment) != len(sharding_device_assignment):
+    raise NotImplementedError(
+      f"Exported module {exported.fun_name} was lowered for "
+      f"{len(sharding_device_assignment)} devices and is called in a context with "
+      f"{len(ctx_device_assignment)} devices"
+    )
+  return mlir.wrap_with_sharding_op(ctx, x, x_aval,
+                                    x_sharding._hlo_sharding.to_proto())
```

Comment:
I would do: `x_sharding._to_xla_hlo_sharding(x_aval.ndim).to_proto()`

Reply:
I believe you that this results in the same result, but why is it better?

There is one way in which the proposed code is not as clear: since the proposed code also depends also on `x_aval.ndim` it is not as obvious that we use the same sharding that was obtained at export-time. The code I have now depends only on `x_sharding`. In reality `x_aval` should be the same aval that was used at export time (this is checked in the abstract eval rule for call_exported), but this requires some thinking.

Reply:
I think you are suggesting this change because it would work for other kinds of sharding objects. Is that it?

Reply:
Yeah, if you think x_aval won't be correct, then there is a bigger problem than using it's ndim to lower to HloSharding.

But maybe as you said, it should be fine. You can add tests to check the avals too (don't know if you already have).